nvflare/private/fed/client/communicator.py

```diff
@@ -177,39 +100,52 @@ def client_registration(self, client_name, servers, project_name):
         """
         local_ip = _get_client_ip()
 
-        login_message = fed_msg.ClientLogin(client_name=client_name, client_ip=local_ip)
-        login_message.meta.project.name = project_name
-
-        with self.set_up_channel(servers[project_name]) as channel:
-            stub = fed_service.FederatedTrainingStub(channel)
-            while True:
-                try:
-                    result = stub.Register(login_message)
-                    token = result.token
-                    ssid = result.ssid
-                    self.should_stop = False
-                    break
-                except grpc.RpcError as grpc_error:
-                    self.grpc_error_handler(
-                        servers[project_name],
-                        grpc_error,
-                        "client_registration",
-                        verbose=self.verbose,
-                    )
-                    excep = FLCommunicationError(grpc_error, "grpc_error:client_registration")
-                    if isinstance(grpc_error, grpc.Call):
-                        status_code = grpc_error.code()
-                        if grpc.StatusCode.UNAUTHENTICATED == status_code:
-                            raise excep
+        login_message = new_cell_message(
+            {
+                CellMessageHeaderKeys.CLIENT_NAME: client_name,
+                CellMessageHeaderKeys.CLIENT_IP: local_ip,
+                CellMessageHeaderKeys.PROJECT_NAME: project_name,
+            }
+        )
+
+        start = time.time()
+        while not self.cell:
+            self.logger.info("Waiting for the client cell to be created.")
+            if time.time() - start > 15.0:
+                raise RuntimeError("Client cell could not be created. Failed to login the client.")
+            time.sleep(0.5)
+
+        while not self.cell.is_cell_connected(FQCN.ROOT_SERVER):
+            time.sleep(0.1)
+            if time.time() - start > 30.0:
+                raise FLCommunicationError("error:Could not connect to the server for client_registration.")
+
+        while True:
+            try:
+                result = self.cell.send_request(
+                    target=FQCN.ROOT_SERVER,
+                    channel=CellChannel.SERVER_MAIN,
+                    topic=CellChannelTopic.Register,
+                    request=login_message,
+                )
+                return_code = result.get_header(MessageHeaderKey.RETURN_CODE)
+                if return_code == ReturnCode.UNAUTHENTICATED:
+                    unauthenticated = result.get_header(MessageHeaderKey.ERROR)
+                    raise FLCommunicationError("error:client_registration " + unauthenticated)
+
+                token = result.get_header(CellMessageHeaderKeys.TOKEN)
+                ssid = result.get_header(CellMessageHeaderKeys.SSID)
+                if not token and not self.should_stop:
                     time.sleep(5)
```

Comment:
This wait is quite long.

Reply:
This is for client registration. I will change it to a little faster.